### ISSUE TITLE
Need ability to commit messages for a particular file in "combine commits for review" mode

### ISSUE THREAD

**Issue by apinstein** [Reviewable/Reviewable#228]
I had an idea when reviewing a large-ish PR today; it would be really cool to see the comment log for a particular file. For instance, when a PR touches 20 files, it can be tough to understand the context of changes that were made on a particular file, the order they were made in, and why. Seeing the relevant commit messages as a reference for each file could be super-helpful. `git log --oneline rN..rM -- path/to/file.php` where `rN` and `rM` are the base/final for that particular file.

Presently this works if you're in *review each commit separately mode* via the "Changes" dropdown. However it doesn't work in the awesome *combine commits for review* mode.

I have set up a sample repo to demonstrate this here: https://reviewable.io/reviews/apinstein/reviewable-demo/1

Something as simple as an affordance in each "file block" to display this info would be a great short-term solution if something more integrated w/overall UX is more complex.

**Comment by pkaminski** [comment]
So there are a couple small bugs in the code that decides which commits to show and which file is current, but those would be relatively simple to fix.  The much bigger problem is that the only way to figure out the files affected by a commit is to compare the git trees, which I currently only do at the revision level to save on processing (some of these trees are huge!).  This means that in "combine commits" mode, where a revision can cover multiple commits, I'll only get attribution for the group of commits as a whole rather than individual ones.

I'll try to see if I can be smarter with fetching and caching the trees so that I can afford to get finer-grained data on the server, and find a compact way to store it so as not to blow out my storage requirements.  Another alternative would be to fetch and diff the trees on demand on the client and take a latency hit (plus a bunch of brand-new code and higher API quota usage).  Either way, this is not going to be a simple fix so I'm afraid I'm going to have to leave it 'til later.  Sorry.

**Comment by apinstein** [comment]
I just looked quickly at the Github API and I suppose I see that you could be doing Reviewable w/o ever cloning the actual git repo (my prior assumption) and just using the API 100%.

If true, I can grok this now:

> I'll try to see if I can be smarter with fetching and caching the trees so that I can afford to get finer-grained data on the server, and find a compact way to store it so as not to blow out my storage requirements. Another alternative would be to fetch and diff the trees on demand on the client and take a latency hit (plus a bunch of brand-new code and higher API quota usage).

I can appreciate both of those issues. As a paying customer this is another good time-saver and would be worth a little more $ if that's what it takes to get more API limit and cover storage.

I'd think caching the file-level commit histories when you create the "batch of commits" is the way to go. Seems quite straightforward to generate and access that data structure from the looking at the API docs.

Hoping for sooner rather than later but I get it :)

Good luck.

**Comment by pkaminski** [comment]
Yeah, there's no repo cloning going on -- in fact, my servers never see customers' code.  And unfortunately the quota in question is yours rather than mine (i.e., it's per GitHub user account, not per app) and there's no way to increase it even with money on the table (I asked).

But I think I can probably do this client-side on demand without too much latency, with a number of requests at worst proportional to the depth of the file in the hierarchy times the number of commits to check.  And since the trees are immutable, they cache really well which will short-circuit a lot of the fetches.  Maybe do some prefetching if needed.  We'll see...
